File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.c
struct ListNode* mergeTwoLists(struct ListNode* h1, struct ListNode* h2) {
    if (h1 == NULL) return h2;
    if (h2 == NULL) return h1;

    if (h1->val <= h2->val) {
        h1->next = mergeTwoLists(h1->next, h2);
        return h1;
    }

    h2->next = mergeTwoLists(h1, h2->next);
    return h2;
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {

        ListNode dummy = new ListNode(-1);
        ListNode ansTail = dummy;

        while (list1 != null && list2 != null) {

            if (list1.val < list2.val) {
                ansTail.next = list1;
                list1 = list1.next;
            } else {
                ansTail.next = list2;
                list2 = list2.next;
            }

            ansTail = ansTail.next;
        }

        if (list1 != null) {
            ansTail.next = list1;
        } else {
            ansTail.next = list2;
        }

        return dummy.next;
    }
}